content/en/docs/Architecture/cloud-init.md

```diff
@@ -41,32 +41,51 @@ stages:
+| **Stage**                  | **Description**                                                                                                                                                                                                                                                                     |
+|----------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| _rootfs_                   | This is the earliest stage, running before switching root, just right after the root is mounted in /sysroot and before applying the immutable rootfs configuration. This stage is executed over initrd root, no chroot is applied.                                                  |
+| _initramfs_                | This is still an early stage, running before switching root. Here you can apply radical changes to the booting setup of Elemental. Despite this is executed before switching root this execution runs chrooted into the target root after the immutable rootfs is set up and ready. |
+| _boot_                     | This stage is executed after initramfs has switched root, during the systemd bootup process.                                                                                                                                                                                        |
+| _fs_                       | This stage is executed when fs is mounted and is guaranteed to have access to the state and persistent partitions ( `COS_STATE`  and  `COS_PERSISTENT` respectively).                                                                                                               |
+| _network_                  | This stage is executed when network is available                                                                                                                                                                                                                                    |
+| _reconcile_                | This stage is executed 5m after boot and periodically each 60m.                                                                                                                                                                                                                     |
+| _kairos-install.pre_       | This stage is executed before installation of the OS starts                                                                                                                                                                                                                         |
```

Comment:
Strange that we have this, and also `_before-install_` I think I'd rather just have his one, and call the other one `_partitioning.pre_`. Is that possible? Same for `_after-install_` it sounds like it should be `_encryption.after_` ... same thought for all the others that are called `_before` or `_after` ... anyways that is a comment for outside this PR but if you agree maybe we can create a ticket

Reply:
yes, This needs a huge overhaul to first be on sync with all actions (like there should be pre and post in all of them and have more and more different stages that we can run. like `install-partition-before` `install-partition-after` `install-encryption-before` and so on.

In other places like reset we should also have things like before oem partitioning in case you want to save something from oem, etc..

Its clear this has gotten a bit abandoned and can be much more specific, much more in line with each action and much richer in opportunities to have the users affect the outcome of them.